src/Optimization/hiopKKTLinSys.hpp

```diff
@@ -524,6 +561,254 @@ class hiopKKTLinSysFull: public hiopKKTLinSysCurvCheck
+  }
+  
+  bool split_x_to_build_it(const hiopVector& x);
+  bool combine_res_to_build_y(hiopVector& y);
+
+  hiopVector* dx_;
+  hiopVector* dd_;
+  hiopVector* dyc_;
+  hiopVector* dyd_;
+  hiopVector* dsxl_;
+  hiopVector* dsxu_;
+  hiopVector* dsdl_;
+  hiopVector* dsdu_;
+  hiopVector* dzl_;
+  hiopVector* dzu_;
+  hiopVector* dvl_;
+  hiopVector* dvu_;
+  
+  hiopVector* yrx_;
+  hiopVector* yrd_;
+  hiopVector* yryc_;
+  hiopVector* yryd_;
+  hiopVector* yrsxl_;
+  hiopVector* yrsxu_;
+  hiopVector* yrsdl_;
+  hiopVector* yrsdu_;
+  hiopVector* yrzl_;
+  hiopVector* yrzu_;
+  hiopVector* yrvl_;
+  hiopVector* yrvu_;
+};
+
+/** 
+ * operators for compressed KKT preconditioner
+ */
+class hiopCompressPrecondOpr : public hiopLinearOperator
+{
+public:
+  hiopCompressPrecondOpr(hiopKKTLinSys* kkt, const hiopIterate* iter, const hiopResidual* resid, const hiopIterate* dir)
+    : kkt_(kkt),
+      iter_(iter),
+      resid_(nullptr),
+      dir_(nullptr)
+  {
+    resid_ = new hiopResidual(kkt_->nlp_);
+    dir_ = new hiopIterate(kkt_->nlp_);
+    dx_ = dir_->get_x();
+    dx_->setToZero();
+    dd_ = dir_->get_d();
+    dd_->setToZero();
+    dyc_ = dir_->get_yc();
+    dyc_->setToZero();
+    dyd_ = dir_->get_yd();
+    dyd_->setToZero();
+    dsxl_ = dir_->get_sxl();
+    dsxl_->setToZero();
+    dsxu_ = dir_->get_sxu();
+    dsxu_->setToZero();
+    dsdl_ = dir_->get_sdl();
+    dsdl_->setToZero();
+    dsdu_ = dir_->get_sdu();
+    dsdu_->setToZero();
+    dzl_ = dir_->get_zl();
+    dzl_->setToZero();
+    dzu_ = dir_->get_zu();
+    dzu_->setToZero();
+    dvl_ = dir_->get_vl();
+    dvl_->setToZero();
+    dvu_ = dir_->get_vu();
+    dvu_->setToZero();
+
+    yrx_ = resid_->get_rx();
+    yrx_->setToZero();
+    yrd_ = resid_->get_rd();
+    yrd_->setToZero();
+    yryc_ = resid_->get_ryc();
+    yryc_->setToZero();
+    yryd_ = resid_->get_ryd();
+    yryd_->setToZero();
+    yrsxl_ = resid_->get_rxl();
+    yrsxl_->setToZero();
+    yrsxu_ = resid_->get_rxu();
+    yrsxu_->setToZero();
+    yrsdl_ = resid_->get_rdl();
+    yrsdl_->setToZero();
+    yrsdu_ = resid_->get_rdu();
+    yrsdu_->setToZero();
+    yrzl_ = resid_->get_rszl();
+    yrzl_->setToZero();
+    yrzu_ = resid_->get_rszu();
+    yrzu_->setToZero();
+    yrvl_ = resid_->get_rsvl();
+    yrvl_->setToZero();
+    yrvu_ = resid_->get_rsvu();
+    yrvu_->setToZero();
+  }
+
+  virtual ~hiopCompressPrecondOpr()
+  {
+    delete resid_;  
+    delete dir_;  
+  };
+
+  /** y = KKT * x */
+  virtual bool times_vec(hiopVector& y, const hiopVector& x);
+
+  /** y = KKT' * x */
+  virtual bool trans_times_vec(hiopVector& y, const hiopVector& x);
+
+private:
+  hiopKKTLinSys* kkt_;
+  const hiopIterate* iter_;
+  hiopResidual* resid_;
+  hiopIterate* dir_;
+
+  hiopCompressPrecondOpr()
+    : kkt_(nullptr),
+      resid_(nullptr),
+      dir_(nullptr)
+  {
+    assert(false && "this constructor should not be used");
+  }
+  
+  bool split_x_to_build_it(const hiopVector& x);
```

Comment:
document these four private functions since they will be revisited in the near future. (issue #423)